File: Collision_Approx_by_Ivan/Python Implementation/Body.py
from typing import Self

class Body:
    def __init__(self, mass:float, side:float, position:float, velocity:float) -> None:
        """
        Handles all the logic for a body for calculating PI
        """        
        self.mass = mass
        self.side = side
        self.position = position
        self.velocity = velocity
        # initializing the parameters 

    def update(self, deltaTime:float) -> None:
        """updating position and handling wall collision"""
        self.position += self.velocity * deltaTime
        self.collideWallHandle()

    @property
    def right(self):
        return self.position + self.side

    def isColliding(self, other:Self) -> bool:
         """
         checks if the body is colliding with any other body
         """ 
         return other.position+other.side > self.position and self.position+self.side > other.position

    def isCollidingWall(self) -> bool:
        """
        chekcing if the body is colliding with the wall
        """ 
        return self.position <= 0

    def collideWallHandle(self) -> None:
        """
        Negating the velocity if the body is colliding with wall 
        """
        
        self.velocity *= -1 if self.isCollidingWall() else 1

    def collisionHandle(self, other:Self) -> None:
        """
        Handles collision with the other body
        """
        
        if self.isColliding(other):
            velocity = (((self.velocity*(self.mass-other.mass)) + (2*other.mass*other.velocity)) / (self.mass + other.mass))
            # velocity equation is taken from the README.md file in the root directory
            
            otherVelocity = (((other.velocity*(other.mass-self.mass)) + (2*self.mass*self.velocity)) / (other.mass + self.mass))
            # velocity equation is taken from the README.md file in the root directory

            self.velocity = velocity
            other.velocity = otherVelocity
            # updating velocity   

File: Collision_Approx_by_Ivan/p5 Implementation/body.js
class Body{
    constructor(mass, side, position, velocity){
        this.mass = mass;
        this.side = side;
        this.position = position;
        this.velocity = velocity;
        // initializing the parameters 
    }

    update(){
        this.position += this.velocity;
        this.collideWallHandle();
        // updating position and handling wall collision 
    }

    isColliding(other){
        return other.position+other.side > this.position && this.position+this.side > other.position;
        // checks if the body is colliding with any other body 
    }

    isCollidingWall(){
        return this.position <= 0;
        // chekcing if the body is colliding with the wall 
    }

    collideWallHandle(){
        if (this.isCollidingWall()) {
            this.velocity *= -1;
            // this negates the velocity 
        }
        // handling collision with the wall
    }

    collisionHandle(other){
        if(this.isColliding(other)){
            const velocity = (((this.velocity*(this.mass-other.mass)) + (2*other.mass*other.velocity)) / (this.mass + other.mass))
            // velocity equation is taken from the README.md file in the root directory
            
            const otherVelocity = (((other.velocity*(other.mass-this.mass)) + (2*this.mass*this.velocity)) / (other.mass + this.mass))
            // velocity equation is taken from the README.md file in the root directory

            this.velocity = velocity;
            other.velocity = otherVelocity;
            // updating velocity 
        }
        // handling the collision 
    }

    render(){
        square(this.position, height-this.side, this.side);
        // rendering the square 
    }
}